### PR TITLE
Added notificationKey and notificationPayload in options

### PR DESCRIPTION
**Pull request by MSkred**

### CODE REVIEW

**Comment by MSkred** [comment]
I mooved the publicVapidKey in the `firebase-messaging-sw.js. Said me if my update is not what do you want, I can update this to day !

**Comment by MSkred** [comment]
I don't know if it possible but we could add a feature to reduce the package size and install only the part the user wants. _example: only messaging if use firebase-module only for push notification_

**Comment by lupas** [comment]
> I mooved the publicVapidKey in the `firebase-messaging-sw.js. Said me if my update is not what do you want, I can update this to day !

What you did is not going to work at all. All I meant is to move the config attributes to `options.services.messaging`, not move the entire messaging module logic into the service worker. That does not work at all.

I'll leave comments in the changed files.

> @lupas I don't know if it possible but we could add a feature to reduce the package size and install only the part the user wants. example: only messaging if use firebase-module only for push notification

This is already implemented. Only services that are setup in the nuxt/firebase config are loaded.
Have a look at [this](https://firebase.nuxtjs.org/guide/getting-started/#simple-configuration), in this example only firebase Auth is imported and initialized.

File: lib/templates/firebase-messaging-sw.js [hunk 1]
```diff
@@ -18,16 +18,39 @@ else {
   firebase.initializeApp(options.config)
 } 
 
+const fireMess = firebase.messaging()
```
Comment (lupas) [inline lib/templates/firebase-messaging-sw.js:21]: why did you move this? This should go back to main.js

Comment (lupas) [reply lib/templates/firebase-messaging-sw.js:21]: (I mean the entire block you copied)

File: docs/guide/getting-started/README.md [hunk 2]
```diff
@@ -122,7 +122,8 @@ modules: [
             }
           },
           messaging: {
-            createServiceWorker: true
+            createServiceWorker: true,
+            fcmPublicVapidKey: '<publicVapidKey>' // OPTIONAL : Sets vapid key for FCM after initialization
```
Comment (lupas) [inline docs/guide/getting-started/README.md:126]: here should also go the `notificationKey` and the `notificationOptionsTemplate `

File: docs/guide/options/README.md [hunk 3]
```diff
@@ -485,7 +483,8 @@ messaging: true
 // or
 
 messaging: {
-  createServiceWorker: false
+  createServiceWorker: false,
+  fcmPublicVapidKey: '<publicVapidKey>' // OPTIONAL : Sets vapid key for FCM after initialization
```
Comment (lupas) [inline docs/guide/options/README.md:487]: here should also go the `notificationKey` and the `notificationOptionsTemplate `

File: lib/plugins/main.js [hunk 4]
```diff
@@ -190,30 +190,6 @@ export default async ({ res }, inject) => {
 
   <% } %>
 
-  /** --------------------------------------------------------------------------------------------- **/
```
Comment (lupas) [inline lib/plugins/main.js:193]: All this needs to stay here in this file since it creates the service worker in the first place.

File: lib/module.js [hunk 5]
```diff
@@ -211,6 +211,8 @@ function validateConfigKeys (options, currentEnv) {
     'appId'
     // 'measurementId' - not a must without Analytics, we throw a warning below
     // 'fcmPublicVapidKey' - not required
+    // 'notificationKey' - not required
```
Comment (lupas) [inline lib/module.js:214]: delete this

File: lib/module.js [hunk 6]
```diff
@@ -211,6 +211,8 @@ function validateConfigKeys (options, currentEnv) {
...
+    // 'notificationPayload' - not required
```
Comment (lupas) [inline lib/module.js:215]: delete this

File: lib/module.js [hunk 7]
```diff
@@ -211,6 +211,8 @@ function validateConfigKeys (options, currentEnv) {
     'appId'
     // 'measurementId' - not a must without Analytics, we throw a warning below
     // 'fcmPublicVapidKey' - not required
```
Comment (lupas) [inline lib/module.js:213]: delete this

**Comment by MSkred** [comment]
Sorry for the mistake, it's my 1st time :) now I think it's what do you want

**Comment by lupas** [comment]
> Sorry for the mistake, it's my 1st time :) now I think it's what do you want

No worries at all, I'm happy if people contribute! The structure on how the module works is not instantly clear, so I totally understand ;-)

It looks much better now, thanks! I will have a look at it and test it later today, will let you know!

**Comment by MSkred** [comment]
> This is already implemented. Only services that are setup in the nuxt/firebase config are loaded.
Have a look at this, in this example only firebase Auth is imported and initialized.

I wanted to talk about node_modules because the size of @firebase and firebase-module are huge for production for example to upload to lambda. As firebase-module needs @firebase the two combined make 83MB (52MB @firebase & 31MB firebase-module)

**Comment by lupas** [comment]
HeadsUp: Haven't checked & merged yet because https://github.com/nuxt-community/firebase-module/pull/174 will change a lot of the code we're also touching here, will have to merge that one first and then see if it still fits.

I'll have a look at this after the merge of 174. Sorry for the delay.

**Comment by MSkred** [comment]
> HeadsUp: Haven't checked & merged yet because #174 will change a lot of the code we're also touching here, will have to merge that one first and then see if it still fits.
> 
> I'll have a look at this after the merge of 174. Sorry for the delay.

Ok, no problem, thank you for the answer. Do you think you will do this for next monday ?

**Comment by lupas** [comment]
> > HeadsUp: Haven't checked & merged yet because #174 will change a lot of the code we're also touching here, will have to merge that one first and then see if it still fits.
> > I'll have a look at this after the merge of 174. Sorry for the delay.
> 
> Ok, no problem, thank you for the answer. Do you think you will do this for next monday ?

I think so yes, but cannot guarantee.

**Comment by mentAl-maZe** [comment]
Hi @MSkred,
I am a collaborator on the pull request (#174) mentioned above by @lupas.
As the pull request changed most of the templates i took the liberty to create a new one based on those changes and implementing what you were trying to do with this, please close this PR and join the discussion over there.

**Comment by lupas** [comment]
Closing in favour of https://github.com/nuxt-community/firebase-module/pull/186
